volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -961,6 +961,43 @@ def __iter__(self) -> Iterator[interfaces.objects.ObjectInterface]:
         return self.to_list(self.vol.parent.vol.type_name, self.vol.member_name)
 
 
+class hlist_head(objects.StructType, collections.abc.Iterable):
+    def to_list(
+        self,
+        symbol_type: str,
+        member: str,
+    ) -> Iterator[interfaces.objects.ObjectInterface]:
+        """Returns an iterator of the entries in the list.
+
+        This is a doubly linked list; however, it is not circular, so the 'forward' field
+        doesn't make sense. Also, the sentinel concept doesn't make sense here either;
+        unlike list_head, the head and nodes each have their own distinct types. A list_head
+        cannot be a node by itself.
+        - The 'pprev' of the first 'hlist_node' points to the 'hlist_head', not to the last node.
+        - The last element 'next' member is NULL
+
+        Args:
+            symbol_type: Type of the list elements
+            member: Name of the list_head member in the list elements
+
+        Yields:
+            Objects of the type specified via the "symbol_type" argument.
+
+        """
+        vmlinux = linux.LinuxUtilities.get_module_from_volobj_type(self._context, self)
+
+        current = self.first
+        while current and current.is_readable():
+            yield linux.LinuxUtilities.container_of(
+                current, symbol_type, member, vmlinux
+            )
+
+            current = current.next
+
+    def __iter__(self) -> Iterator[interfaces.objects.ObjectInterface]:
+        return self.to_list(self.vol.parent.vol.type_name, self.vol.member_name)
```

Comment:
Is `member_name` defined as part of the type when an `hlist_head` is constructed?  I'm just interested to know what populates this value and what happens if it's not present?  This feels a little brittle at the moment...

Reply:
Unfortunately, it's not. See the contructors here: https://elixir.bootlin.com/linux/v6.11.4/source/include/linux/list.h#L939.
The type is based on the destination variable type in the foreach preprocessor macros here: https://elixir.bootlin.com/linux/v6.11.4/source/include/linux/list.h#L1162

Reply:
Sooooo, how then does it get into the object's `vol` dict?  5:)

Reply:
oh, sorry, I thought you were talking the struct in the kernel. 
Yes, you are right.. good catch! I removed this object's ability to be directly iterable. It needs to be called via `to_list()` to provide the correct member_name.

Reply:
hm I think the same happens with the `list_head` objects. It requires to explicitly set a `member_name`, not sure if it's used directly as an iterator somewhere, maybe that's why it never failed.